politeiad/cache/cockroachdb/decred.go

```diff
@@ -884,6 +884,226 @@ func (d *decred) cmdTokenInventory(payload string) (string, error) {
+		return startVotes, nil
+	}
+
+	tokens := make([]string, 0, len(authorizeVotes))
+	for token := range authorizeVotes {
+		tokens = append(tokens, token)
+	}
+
+	svs := make([]StartVote, 0, len(tokens))
+	err := d.recordsdb.
+		Where("token IN (?)", tokens).
+		Preload("Options").
+		Find(&svs).
+		Error
+
+	if err != nil {
+		return nil, err
+	}
+	for _, sv := range svs {
+		startVotes[sv.Token] = sv
+	}
+
+	return startVotes, nil
+}
+
+// lookupResultsForVoteOptions looks in the CastVote to see how many votes
+// each option has received.
+func (d *decred) lookupResultsForVoteOptions(options []VoteOption) ([]decredplugin.VoteOptionResult, error) {
+
+	results := make([]decredplugin.VoteOptionResult, 0, len(options))
+
+	for _, v := range options {
+		var votes uint64
+		tokenVoteBit := v.Token + strconv.FormatUint(v.Bits, 16)
+		err := d.recordsdb.
+			Model(&CastVote{}).
+			Where("token_vote_bit = ?", tokenVoteBit).
+			Count(&votes).
+			Error
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results,
+			decredplugin.VoteOptionResult{
+				ID:          v.ID,
+				Description: v.Description,
+				Bits:        v.Bits,
+				Votes:       votes,
+			})
+	}
+
+	return results, nil
+}
+
+// getVoteResultsForStartVotes retrieves vote results for records that have
+// begun the voting process. Results are lazily loaded into this table, so
+// some results are manually looked up in the CastVote table.
+func (d *decred) getVoteResults(startVotes map[string]StartVote) (map[string][]decredplugin.VoteOptionResult, error) {
+
+	results := make(map[string][]decredplugin.VoteOptionResult)
+
+	if len(startVotes) == 0 {
+		return results, nil
+	}
+
+	tokens := make([]string, 0, len(startVotes))
+	for token := range startVotes {
+		tokens = append(tokens, token)
+	}
+
+	vrs := make([]VoteResults, 0, len(tokens))
+	err := d.recordsdb.
+		Where("token IN (?)", tokens).
+		Preload("Results").
+		Preload("Results.Option").
+		Find(&vrs).
+		Error
+	if err != nil {
+		return nil, err
+	}
+
+	for _, vr := range vrs {
+		results[vr.Token] = convertVoteOptionResultsToDecred(vr.Results)
+	}
+
+	for token, sv := range startVotes {
+		_, ok := results[token]
+		if ok {
+			continue
+		}
+
+		res, err := d.lookupResultsForVoteOptions(sv.Options)
+		if err != nil {
+			return nil, err
+		}
+
+		results[token] = res
+	}
+
+	return results, nil
+}
+
+func (d *decred) cmdBatchVoteSummary(payload string) (string, error) {
+	log.Tracef("cmdBatchVoteSummary")
+
+	bvs, err := decredplugin.DecodeBatchVoteSummary([]byte(payload))
+	if err != nil {
+		return "", err
+	}
+
+	// This query gets the latest version of each record
+	query := `SELECT a.* FROM records a
+	LEFT OUTER JOIN records b
+		ON a.token = b.token AND a.version < b.version
+	WHERE b.token IS NULL AND a.token IN (?)`
+
+	rows, err := d.recordsdb.Raw(query, bvs.Tokens).Rows()
+
+	if err != nil {
+		return "", err
+	}
+	defer rows.Close()
+
+	records := make(map[string]Record)
```

Comment:
Map and slice sizes should be pre allocated when you know the desired capacity.
```suggestion
	records := make(map[string]Record, len(bvs.Tokens))
```

Same for `summaries`.